floor_generator/lib/processor/query_processor.dart

```diff
@@ -0,0 +1,145 @@
+import 'package:floor_generator/misc/extension/dart_type_extension.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:floor_generator/processor/error/query_processor_error.dart';
+import 'package:floor_generator/processor/processor.dart';
+import 'package:floor_generator/value_object/query.dart';
+
+class QueryProcessor extends Processor<Query> {
+  final QueryProcessorError _processorError;
+
+  final String _query;
+
+  final List<ParameterElement> _parameters;
+
+  QueryProcessor(MethodElement methodElement, this._query)
+      : _parameters = methodElement.parameters,
+        _processorError = QueryProcessorError(methodElement);
+
+  @override
+  Query process() {
+    _assertNoNullableParameters();
+
+    final listParameters = <ListParameter>[];
+    final newQuery = _processParameters(listParameters);
+
+    return Query(
+      newQuery,
+      listParameters,
+    );
+  }
+
+  String _processParameters(List<ListParameter> listParametersOutput) {
+    final indices = <String, int>{};
+    final fixedParameters = <String>{};
+    //map parameters to index (1-based) or 0 (if its a list)
+    int currentIndex = 1;
+    for (final parameter in _parameters) {
+      if (parameter.type.isDartCoreList) {
+        indices[':${parameter.name}'] = 0;
+      } else {
+        fixedParameters.add(parameter.name);
+        indices[':${parameter.name}'] = currentIndex++;
+      }
+    }
+
+    //get List of query variables
+    final variables = findVariables(_query);
+    _assertAllParametersAreUsed(variables);
+    final newQuery = StringBuffer();
+    int currentLast = 0;
+    for (final varToken in variables) {
+      newQuery.write(_query
+          .substring(currentLast, varToken.startPosition)
+          .replaceAll('\n', ' '));
+      final varIndexInMethod = indices[varToken.name];
+      if (varIndexInMethod == null) {
+        throw _processorError.unknownQueryVariable(varToken.name);
+      } else if (varIndexInMethod > 0) {
+        //normal variable/parameter
+        if (varToken.isListVar)
+          throw _processorError.queryArgumentsAndMethodParametersDoNotMatch;
+        newQuery.write('?');
+        newQuery.write(varIndexInMethod);
+      } else {
+        //list variable/parameter
+        if (!varToken.isListVar)
+          throw _processorError.queryArgumentsAndMethodParametersDoNotMatch;
```

Comment:
See comment above ☝️

Reply:
this happens if the parameter is a List and the query location looks like it shouldn't accept a list (and the reverse for above). There also is a test for that and the case above.

I agree though, the error message(s) could be improved.

Reply:
error messages were improved.